Rewrite, improve, update and correct the documentation

Korne127 opened the pull request:
This is still a work in progress. However, I now got most of the major changes done, and since I might still take a while for the rest and am a bit afraid that someone else might want to start doing the same, I decided to open the current state as a draft pull request.

----

When I started looking into how to deploy AppImages a few weeks ago, I had big troubles understanding the documentation:
- There was *very* much duplicated information which made it hard to find a specific detail that was only mentioned in one of the duplicated parts and generally to read through the documentation.
- There was *very* much outdated information, from commonly recommending the deprecated linuxdeployqt tool, to the deprecated AppRun.c file, not even mentioning the new go-appimagetool, many invalid and outdated links, old Linux versions (one even being eleven years (!) old), etc.
- The same terms were sometimes used for very different concepts, e.g. "manual packaging" for
  1. manually creating the *entire* AppDir that's then converted into an AppImage by appimagetool,
  2. manually creating the AppDir structure to use with higher level tools that include the shared libraries *and*
  3. giving the path of executables, desktop and icon files to linuxdeploy via command line arguments
  
  which obviously led to much confusion.
- There was much conflicting information, mostly due to duplicated information, which has been outdated in some parts and updated in others, or the same terms used for different concepts.
- There was missing information that was expected in other places, e.g. which resolutions are valid for icon files.
- There was information scattered across places or in the wrong places, e.g. several pages contained some details about desktop files, but no page in the packaging guide clearly explained everything related to it and how to create them.
- There were weird structure decisions, e.g. placing a software overview, mostly about low-level software that the average user should not use, in the introduction section instead of the reference section.
- There was no single detailed comparison about the differences between the individual tools and approaches to create an AppImage, resulting in a lot of confusion which one(s) to use and how to create AppImages.
- Some things were just explained badly, e.g. the pages about using make to create an AppDir lacked the most important information on how to do so, and the linuxdeploy page made it very hard to understand how to use it without make.

I think much of this is not that noticeable when you know your way around AppImages well enough. Sure, you might notice many things aren't up to date and that some things are duplicated, but I think you can generally read through it, understand what's meant, fill in the missing information and don't really realize how hard it is to understand all of that if you don't have that prior experience and knowledge.

I also think that much of this comes from the fact that the AppImage ecosystem has evolved a lot in the past years. There are many things that have considerably improved, e.g. going from having to package the entire AppDir by yourself and use appimagetool directly, over tools like linuxdeployqt that package all shared libraries from an existing AppDir structure, to linuxdeploy, which can get all necessary data with command line arguments and requires no pre-existing structure.
I also noticed that some old parts were expecting the reader to repackage existing packages, probably out of a time where AppImages haven't been commonly used, while other parts were mainly directed to the application authors.
I can imagine that the focus was on developing better solutions and improving the AppImage packager and user experience. And it's understandable that, with many parts duplicated, you don't always find the time to really go through the documentation and update everything correctly when something changes, slowly leading to a worse documentation state.

All of this is to say, I hope that you don't take this as an insult. I do appreciate the work you put into AppImages and the whole ecosystem and I understand that the focus has been on other things, but I do think that the documentation is in a bad state.

And this also led to frustration by other users. See for example https://discourse.appimage.org/t/does-appimage-include-all-shared-library-dependencies-by-default/509/6, where someone is frustrated with the documentation as they don't find a guide on how to automatically package shared libraries, or https://github.com/ruffle-rs/ruffle/issues/12401, where someone didn't even know about the other available tools that help with this, which is the core thing the documentation should explain in my opinion.

----

Therefore I started to rewrite, improve, update and correct the documentation.
I started with the linuxdeploy page, which I had personally had a lot of struggle with to understand previously. From there on, I worked my way through the documentation (mostly the packaging guide), and went on to improve the structure, add a comparison of creation tools & approaches, deliver better explanations, update the outdated information and add additional information.

Here is an overview of my changes:

### New sections that contain collected information
- One unified section for all AppImage creation tools has been added.
  - Its main page contains one central table which lists the advantages, disadvantages and differences between the different methods.
  - It contains one page for each AppImage creation tool (some of which have been collected from many places throughout the documentation), explaining how to use it.
    - This also means that a new page for go-appimagetool has been added (it had previously not been mentioned).
    - Much additional information and better explanations have also been added to most of these pages.
    - The linuxdeploy page has been rewritten, improved and updated. It’s now much easier to understand how to use it.
  - The section also contains a better overview on what the AppImage creation tools actually do, which helps to understand why they should be used.
- The information about the different packaging approaches (packaging as application author and repackaging existing binaries) has been collected in one section and rewritten, with additional information and a better explanation of the differences between the approaches.
- One unified section on how to create the AppDir structure has been added.
  - Its main page provides an overview of the different ways to do it, the reason why one would want to do it and the additional possibility of manually packaging everything.
  - This also means that the concepts of manually creating the AppDir structure (to give it to a tool) and manually packaging everything (without a tool) have been separated and the difference is well-explained.
- A new page about desktop entry files and icon files has been created. Information from several files has been merged together with much additional information.

### Other sections that have been rewritten
- The packaging automation section (formerly called hosted services) has been rewritten and updated.
  - Its main page now contains a new introduction about the advantages of packaging automation and why it should be used.
  - A new tutorial on GitHub Actions has been added to it.
  - The existing (outdated) parts have been updated.
- The sections about which dependencies are bundled and which Linux distribution version the binaries should be compiled on in the concept page have been partially rewritten.
  - Misleading headlines have been changed and a better explanation of why AppImages should usually be compiled on older systems, and which exceptions and other options exist, has been added.
- The packaging guide introduction has been rewritten to give a better introduction of what the guide explains.

### Updated & improved sections
- The AppDir specification has been updated.
  - Information about the AppRun file has been updated and added.
  - More information about the desktop entry and icon files has been added.

### Split up sections
- The software overview section has been split up. Most of it has been moved into the reference section, but some parts moved into other sections in which the respective software had already been mentioned. A better introduction has been added to it as well.

### Minor improvements
- The table formatting has been improved. Tables now have automatic line breaks (instead of scroll bars) and are easier to edit. (This is minor for the existing documentation, but the overview table wouldn’t have been possible without it).
- Many other sections and pages that haven’t been explicitly mentioned have gotten some improvements and additional information.

For more details, each commit contains an in-depth message of all changes I have made within it.

Also, this pull request documents the features introduced in https://github.com/linuxdeploy/linuxdeploy/pull/288, so it depends on that being merged.

----

As I said in the beginning, this is currently a work in progress. I still have a lot of todos and there are quite some sections I haven't really gone through yet at all.
But I still thought that now is a good time to create this draft PR, as most of the major restructuring is done and it's easy to see my desired end result. I'll probably still take a while to finish this (especially as I basically worked the entire last week from waking up until going to bed on this, and will continue slower as it's now known that I'm working on it).

Samueru-sama commented:
I support this, recently there was this discussion asking about the static and deploy everything tools: https://github.com/orgs/AppImage/discussions/1343

Btw a small nitpick, the part that says: 

`can also include these expected core libraries. This considerably increases the AppImage size (by at least 30MB)`

Can be changed to say instead `by at least 10 MiB` and I've found cases where it is even less than that.

Korne127 commented:
@Samueru-sama I'm sorry for the late answer, I was really busy the last days.
Thanks for the notice! I'll definitely update the documentation about the new / different runtimes, especially as currently, all information about the different runtimes is not well documented and rather hard to understand. As I said, this is still a work in progress, and while I took a pause, I'll continue on this soon. Thanks for bringing it to my attention. :)

On the other part, I took that information from the [appimage-builder documentation](https://appimage-builder.readthedocs.io/en/latest/faq.html). But thanks for the feedback; I'll change the part to say that it's usually a big increase, but can also be smaller, depending on which libraries are called.

Korne127 commented:
@probonopd Hey, did you see this PR yet? I noticed that you were also working on the documentation in #103. While luckily, this doesn't interfere with my PR as it's a new section, I would have expected some kind of reaction or response to this.

probonopd commented:
Hi @Korne127, thanks for this massive undertaking.

Hi @TheAssassin is there any way to get this documentation rendered out for easier review (without overwriting the current one, s that I can open them side by side)?

probonopd commented:
Hi @Korne127 

> This considerably increases the AppImage size (by at least 30MB)

Where did you get this number from? If I remember correctly, the results i got were around 1/3rd that number. How did you test?

Korne127 commented:
@probonopd One way to look at the result would be to just clone my version and execute it:
```shell
git clone https://github.com/Korne127/docs.appimage.org.git
cd docs.appimage.org
mkdir venv
python3 -m venv venv
source venv/bin/activate
pip3 install -r requirements.txt
make html
```
And then the files are build in build/html, you just need to double-click them to open in a browser. I also discovered there is a `make.sh` script doing the (I think) five last steps for you.

Korne127 commented:
> Where did you get this number from?

I took them from [the appimage-builder documentation](https://appimage-builder.readthedocs.io/en/latest/faq.html). If the numbers are wrong and your tests consistently produced different results, I'll adapt it.

probonopd commented:
Note to self,

```
sudo apt-get update
sudo apt-get install -y python3-pip python3.10-venv git
git clone https://github.com/Korne127/docs.appimage.org.git
cd docs.appimage.org
mkdir venv
python3 -m venv venv
source venv/bin/activate
pip3 install -r requirements.txt
make html
open html/index.html
```

works nicely for me. Maybe we should even document it ;)

Korne127 commented:
I finally continued on this, and went through the whole introduction and user guide pages and improved, rewrote, updated and corrected them all. I've spent most of my free time in the past week in that process. 😅 (I also merged two pages that basically provided the same kind of information in the same structure with big parts duplicate.) To view the exact changes and things I did, see the commit messages.
I think I'm about done 80%? There is still quite something on my to do list, I've gone through most pages now. Hopefully I'll finish this sometimes soon, because I think this is desperately needed.

Also, I noticed something important: The FUSE page in the AppImage wiki (https://github.com/AppImage/AppImageKit/wiki/FUSE) has generally the same structure behind its documentation page, but has been much more up-to-date (although it also missed some information).
Nevertheless, I think that there are many places in which things are documented: The AppImage wiki, the official documentation, the README files of many projects, other wikis that exist for these projects, etc. And because of that, information just gets updated in one place, which leads to more and more wrong or outdated information piling up.
I think we should place all up-to-date information in one place again (that's essentially what this PR does), and then remove the other wikis / keep their content to a minimum and always link to that one documentation. So when it gets updated, no matter where someone initially searches, they always get the full up-to-date information.

Korne127 commented:
I have another update: I finished rewriting the updating information page: Improving the page structure and all explanations (making them more concise and easier to follow), correcting wrong information that didn't work and adding missing information, giving a better overview about different approaches and their (dis-)advantages and replacing information that depends on a specific project structure with information that's generally applicable.
I also added a corresponding section in the user guide with the basic information a user should know.

Korne127 commented:
Another bulk of changes I've spent the last week working on :D

It's hard to summarise the changes because it's just so much, but I still tried to do it. There are many improvements that aren't included in the following summary; the individual commit messages are much more complete and detailed.

Those were the major improvements of these new changes:
- The whole reference section has been restructured and improved:
   - Content that is directed to users or application authors have been merged into the guides or removed when duplicated.
   - All reference information about AppImages has been rewritten and bundled into a new AppImage reference section. This way, the specification and reference implementation decisions could be separated much more clearly and many duplicates could be removed.
   - Much information (such as about different AppImage versions, their differences and history) has also been added.
   - The whole reference section has been updated accordingly to the change to the new static runtime being the official reference implementation and AppImageKit being split up into different repositories.
- The introductory pages have been rewritten and heavily improved:
   - The front page now directly links to the chapters that people in different situations should read, and makes it much easier to find important information most people might search for.
   - The quickstart and the FAQ have been merged to have a single page to point new users towards, which answers all common questions.
   - The introductory chapter pages have been improved to give a better summary of their content.
- The AppImageHub section has been rewritten to a general software catalog section, giving an overview on what a software catalog is and why one would want to use them and listing the different software catalogs. A corresponding downloads section has been added to the user guide.

Those are other pages that have been rewritten:
- The signature page has been improved, with more concise explanations and an overview as well as information on how to use signatures with the individual tools.
- The AppStream page has been improved, with a new validating section, advantages and a new explanation on how to embed AppStream files with AppImage creation tools that create the AppDir.


And those are general improvements that have been done to the documentation:
- Many pages have been restructured with an improved order and a better explanation and overview on what the page actually explains.
- Duplicated information has been removed and explanations have been made easier to understand with more examples.
- Outdated information, e.g. about the discourse forum, has been removed.
- Wrong links and references have been corrected.

Korne127 commented:
@probonopd I am a bit confused about your comment in https://github.com/AppImage/AppImageKit/issues/877#issuecomment-2564702332.

As I've worked my way through the entire AppImage project while updating this documentation (being it the AppImage creation tools, the updating system, software catalogs or now the history and different types of AppImages), I've made some issues, comments or pull requests on respective repositories. I usually wrote that I stumbled across that while rewriting the documentation.

But now you've stated that I should make an issue before working on the documentation, implying you are not aware of my basically complete rewrite of it (which is really necessary, see the motivation I wrote when I made this PR).
So yeah, I am a bit confused, and as I've spent several months working on this (and far over a hundred working hours), of course also a bit anxious you might have different plans.

Therefore I wanted to ping you again and hurried the last week to finish this batch of changes (as this PR is now mostly done, there are only many minor changes and a few pages I want to majorly improve left).

The work required to rewrite the documentation to be consistently up-to-date without duplications with clearly separated chapters so that every specific information is easy to find and intuitive and concise explanations and information is insane, so I really hope that you can see and appreciate it.

Korne127 commented:
Another smaller update; I finally finished going through every page once. There are still some things on my to do list, but most are either smaller consistency changes or things I already wrote a todo for.

Those are the new changes:
- The testing page has been majorly improved, updated and corrected.
    - The testappimage part has removed (as it's outdated and doesn't work with modern AppImages) and replaced by an explanation on how to manually test with live images.
    - The appimage-testsuite part has been removed (as it's outdated and doesn't work out of the box) and replaced by a new script which is an improved version (which works, is not outdated, future-proof and consists of less code) with explanations.
    - The descriptions and explanations on the page have also been improved.
- Information about the desktop integration tools Gear Lever, zap, Soar and AM has been added to the desktop integration page in the user guide. The other tools' information has also been improved.
- The pkg2appimage page has generally been improved to not have duplicated sections and to have easier to understand explanations and updated to use the prepackaged AppImage.
- The FUSE troubleshooting page has been updated to account for the fact that new static runtime is now the official reference implementation runtime.
- The troubleshooting page has been improved to contain the information on how to see potential error messages.

Korne127 commented:
I also found this comment (https://github.com/AppImage/docs.appimage.org/issues/35#issuecomment-555474893) which I fully agree with.

I think this whole problem that the information was / is inconsistent, outdated, conflicting, etc. stems from the fact that there are several places in which the same kind of information is.
And when things get outdated over time, one thing might be updated here, another thing might get outdated in the wiki, a third thing might get updated in a README, and then you have the mess of different inconsistent and outdated information scattered across the documentation, the AppImageKit wiki, other wikis, the forum, READMEs in different places in different organisations, etc. (and the same thing happened with the duplicated information in the documentation internally).

Therefore, I think that after this documentation is finally updated, structured, easy to understand and without duplications, it should become what it claims to be: **A single unified source of information**.

So after this is merged, I'll go through the wiki and all these other places in the next time and see if there is any additional information and add it to the documentation so that we can remove the wikis and other places and just replace them with links to the documentation, so that everyone always sees the complete unified source of up-to-date information.

freakboy3742 commented:
@Korne127 Thanks for tagging me on this; these changes definitely look extensive.

However, I'm no longer interested in AppImage as a deployment mechanism. The tool I maintain ([Briefcase](https://github.com/beeware/briefcase)) has support for AppImage, but we've [effectively deprecated our support for AppImage](https://briefcase.readthedocs.io/en/stable/reference/platforms/linux/appimage.html) as a result of the large number of issues that we've had with apps deployed using this technique, including (but not limited to) [this one](https://github.com/linuxdeploy/linuxdeploy-plugin-gtk/issues/42). As a result, we now discourage the use of AppImage for new projects, and we plan to eventually remove AppImage support from Briefcase altogether.

Samueru-sama commented:
> @Korne127 Thanks for tagging me on this; these changes definitely look extensive.
> 
> However, I'm no longer interested in AppImage as a deployment mechanism. The tool I maintain ([Briefcase](https://github.com/beeware/briefcase)) has support for AppImage, but we've [effectively deprecated our support for AppImage](https://briefcase.readthedocs.io/en/stable/reference/platforms/linux/appimage.html) as a result of the large number of issues that we've had with apps deployed using this technique, including (but not limited to) [this one](https://github.com/linuxdeploy/linuxdeploy-plugin-gtk/issues/42). As a result, we now discourage the use of AppImage for new projects, and we plan to eventually remove AppImage support from Briefcase altogether.

That webkitgtk is a cursed thing, we had this [issue](https://github.com/VHSgunzo/sharun/issues/2) and we were able to bundle it, but it required a lot of manual hacks. 

The issue is that the library is hardcoded to look for some binaries in `/usr/lib` and it also seems that you need to bundle all the opengl libs as well otherwise you will be getting `EGL_BAD_PARAMETER` errors.

Samueru-sama commented on source/introduction/concepts.rst:
```diff
@@ -14,62 +16,62 @@ The AppImage development follows a few easy-to-understand core principles and co
 .. note::
-
-   On a regular basis, `users ask <https://github.com/AppImage/AppImageKit/issues/848>`__ about implementing support for some sort of "reusable/shared frameworks". These frameworks are supposed to contain bundles of libraries which are needed by more than one AppImage, and hence could save some disk space. For management, they suggest complex automagic systems that will automatically fetch the "f ...
+   On a regular basis, `users ask <https://github.com/AppImage/AppImageKit/issues/848>`__ about implementing support for some sort of "reusable/shared frameworks". These frameworks are supposed to contain bundles of libraries which are needed by more than one AppImage, and hence could save some disk space. For management, they suggest complex automagic systems that will automatically fetch the "f ...
```
>These frameworks are supposed to contain bundles of libraries which are needed by more than one AppImage, and hence could save some disk space.

Likely when this was originally written that was true, [but today that's far from true](https://i.imgur.com/zPRRuKK.png).

I would either remove this or mention that it is not true that they save disk space, but not big deal anyway.

Korne127 replied:
This section isn't about the comparison to a different system like Flatpak, but rather about the general concept of additionally deduplicating AppImage parts. For example, by removing something that's very common from AppImages and instead fetching that once if needed.
The paragraph states that yes, even if those would save disk space (as the AppImages are the same, just with less), that still goes against the core concept of AppImages and would not be worth it.

Samueru-sama commented on source/introduction/concepts.rst:
```diff
@@ -14,62 +16,62 @@ The AppImage development follows a few easy-to-understand core principles and co
+-----------------------------------------------------------
 
-.. _build-on-old-systems:
+AppImages should usually exclude certain "core libraries", which can be expected to be present on all major desktop Linux distributions, reducing the overhead of :ref:`one app = one file <one-app-one-file-principle>`. These dependencies are mostly shared libraries and involve low level libraries like ``libc.so.6`` (the GNU C library, the C language standard library the majority of all Linux distr ...
```
Would put "can" instead of "should" here. I often see people confused when I suggest them to bundle everything to fix an issue and they point to suggestions like this.

Korne127 replied:
As far as I understand it, the main advice is to still use the excludelist, as including some hardware libraries can cause issues or make the AppImage non-working. So ignoring the excludelist is possible, if someone has certain issues, but shouldn't be the default.

But if that information is outdated / doesn't apply anymore, I can update it. @probonopd What do you think about this?

Samueru-sama replied:
@Korne127 probono actually recently used sharun to make this AppImage of [PrusaSlicer](https://github.com/prusa3d/PrusaSlicer/issues/13653) that bundles everything including the gpu libraries.

Samueru-sama commented on source/packaging-guide/appimage-creation-tools/go-appimagetool.rst:
```diff
@@ -0,0 +1,93 @@
+It requires a manual creation of the AppDir folder structure and file placement (if make isn't used). For more information on how to use make accordingly, or manually create the necessary structure, see :ref:`creating-appdir-structure`.
+
+.. note::
+   go-appimagetool is not using the :ref:`reference implementation <reference-implementation>` (called appimagetool) to create AppImages; instead, it creates them itself. This means that it might choose different :ref:`implementation decisions <appimagetool>` (e.g. CLI options like ``--appimage-extract`` the resulting AppImage supports), resulting in AppImages that behave differently to the usual ...
```
Appimages created with go-appimagetool use the same static runtime, maybe the version is a bit older compare to the type2-repo, but the CLI arguments are the same. 

What is different is that go-appimagetool itself has less CLI deploy options than appimagetool, like for example that you can't disable appstream validation.

Korne127 replied:
Ah, thanks for bringing this to my attention! In the reference implementation page, I listed the things both the runtime implementation and the appimagetool implementation determine, and I made a mistake and put the CLI options in the wrong place. That's also the root of this mistake.

I made a [PR](https://github.com/probonopd/go-appimage/pull/326) which makes go-appimagetool use the official reference implementation in the type 2 repository instead of the one it currently uses. After this gets merged, we can be sure that the AppImages behave consistently and that there won't be any difference in the runtime implementations.
I'll then update both pages appropriately (as well as some other parts which mention the differences between the different implementations).

As go-appimagetool still is a different implementation, e.g. with other CLI deploy options (and which could theoretically use a different filesystem), I'll still keep such a disclaimer, but clarify that the runtime is the same.

Korne127 commented:
@freakboy3742 While it's true that there are issues with AppImage and project inactivity, it's sad to hear that you generally discourage its useage, especially as AppImage is a solution in some cases other packaging formats aren't suitable for (e.g. paid software or testing multiple versions side-by-side).

Nevertheless, even if you don't want to actively use AppImage anymore, I'd still appreciate if you could just take a look at this new documentation (the packaging guide) and confirm whether it's now explained in an understandable way how to create AppImages.

probonopd commented:
Thanks @Korne127. I see you are putting a lot of work and effort into this. 
My review is pending on

* https://github.com/AppImage/docs.appimage.org/issues/105
* Then, it would be helpful to have one PR per page or something like that, because I won't find the amount of time to look at all changes at the same time

Again, thank you very much for your dedication, which I appreciate.

Korne127 commented:
It doesn't make sense to split this PR up as it basically is a rewrite of the entire documentation, which does major structure changes to prevent the duplication of many pages and information and produce a better intuitive structure with a single clear place for each information (while previously, some things were completely mixed up without a sensible structure, hidden in some pages, duplicated, etc.).

It definitely doesn't make sense to try to convert the documentation without this PR to Markdown, as it's just… completely outdated, contains much duplicated and wrong information, uses the same terms for different concepts, doesn't explain many important things and misses insanely much information (a big part of this PR is adding much needed previously missing information).

If anything, I could try to convert this to Markdown (either in this PR or in a next one after it gets merged). I personally don't really care (while I think Sphinx may be more professional as most documentations use it, it's very similar: You write text in both and just use other lines for headings instead of using #s, etc.).

But I think that it's really urgent to give this documentation this much-needed update as the current state is serious; just look at the loads of issues this PR fixes with a complete overhaul and at all the changelog what has been improved and fixed.

TheAssassin commented:
As the author of the initial docs, I want to provide some feedback.

First of all, this pull request is pretty much way too large to review. Unfortunately, @Korne127, I'm not aware that you ever tried to contact us beforehand in order to coordinate changes or discuss ideas before spending time on the implementation. People in free software projects discourage this practice because it can lead to useful changes being rejected because they overwhelm the maintainers. And I think we all agree that this is the worst possible outcome. I've had to reject such PRs in the past and I had someone reject my own, so I have some experience here.

The practical issue: I'm not sure I'd ever get around to review this amount of code at once, not to mention back and forth in a review process. Splitting this into some chunks might help reduce the scope to a manageable amount of changes. As I rewrote existing documentation myself beforehand, I know that this is not an easy task, but I believe it might be necessary, especially if you want to speed up processes here. Otherwise, a review would take months without any guarantees on the result.

I've peeked at some changes. Overall, this looks promising. That's all I can say without spending at least half a day on this PR to get a superficial idea.

Regarding @probonopd's comment:

> https://github.com/AppImage/docs.appimage.org/issues/105

That is way out of scope and not at all required for a simple review of the content. Please do not try to enforce your personal agenda with the help of contributors like this. This imposes a very much unnecessary burden. I know you dislike reStructuredText, but if any, this can be discussed separately. The issue even is open and no consensus has been reached.

TheAssassin commented:
P.S.: There is some planning going on in the background which I want to share with you that might influence your writing:

- the static runtime is going to become the default at some point in the future
- AppImageLauncher is on track for a new release 3.0.0, which finally allows it to work with such AppImages, too
- we do have plans for a type 3 and type 4 which are going to make working with AppImages easier, too

TheAssassin commented:
P.P.S.: @Korne127 is there a way to reach you via IRC or Matrix? It might be easier to discuss some of this in some text-based chat or maybe a call.

TheAssassin commented on .woodpecker/docs.yml:
```diff
@@ -13,7 +13,7 @@ pipeline:
       - mkdir -p build/html/download
       - cp build/epub/AppImage.epub build/html/download/"$epub_filename"
-      - bash -xe ci/embed-epub-link.sh "$epub_filename"
+      - bash -xe .woodpecker/embed-epub-link.sh "$epub_filename"
```
We'll most likely switch to a Docker-based deployment workflow in the future. I'll open an issue.

TheAssassin commented on README.rst:
```diff
@@ -1,29 +1,63 @@
-   :target: https://drone.assassinate-you.net/AppImage/docs.appimage.org
-
-.. |ci_status| image:: https://github.com/AppImage/docs.appimage.org/workflows/CI/badge.svg
+.. image:: https://github.com/AppImage/docs.appimage.org/workflows/CI/badge.svg
```
Why remove the alt text?

Korne127 replied:
What alt text did I remove? As far as I can tell, I only removed the CD status badge as it doesn't exist anymore.

TheAssassin commented on README.rst:
```diff
@@ -1,29 +1,63 @@
+   > cd docs.appimage.org
+
+   # Set up and build the documentation
+   > make.sh html
```
```suggestion
   > ./make.sh html
```

Korne127 replied:
Ah, thanks

TheAssassin commented on README.rst:
```diff
@@ -1,29 +1,63 @@
+   # Set up and build the documentation
+   > make.sh html
+
+Calling make.sh will build the documentation after setting up the project and everything required to build the documentation (such as creating a Python virtual environment and installing the dependencies in it). It will only perform the preparation steps that haven't been done before, so you can simply call it each time you want to re-build the documentation.
```
```suggestion
Calling ``make.sh`` will build the documentation after setting up the project and everything required to build the documentation (such as creating a Python virtual environment and installing the dependencies in it). It will only perform the preparation steps that haven't been done before, so you can simply call it each time you want to re-build the documentation.
```

TheAssassin commented on README.rst:
```diff
@@ -1,29 +1,63 @@
+
+Calling make.sh will build the documentation after setting up the project and everything required to build the documentation (such as creating a Python virtual environment and installing the dependencies in it). It will only perform the preparation steps that haven't been done before, so you can simply call it each time you want to re-build the documentation.
+
+**Live reloading:** You can use ``make.sh watch`` instead of ``make.sh html`` |live_reloading|
```
```suggestion
**Live reloading:** You can use ``./make.sh watch`` instead of ``./make.sh html`` |live_reloading|
```

Korne127 replied:
Thanks

TheAssassin commented on README.rst:
```diff
@@ -1,29 +1,63 @@
+
+**Live reloading:** You can use ``make.sh watch`` instead of ``make.sh html`` |live_reloading|
+
+Manually
+++++++++
+
+If you want to set up the project and build the documentation manually, you can use the following commands. (The ``make.sh`` script does essentially the same, just with additional tests whether these preparation steps have already been done before.)
+
+.. code-block:: shell
+
+   # Clone the project
+   > git clone https://github.com/AppImage/docs.appimage.org.git
+   > cd docs.appimage.org
+
+   # Create and activate a Python virtual environment
+   > mkdir venv
+   > python3 -m venv venv
+   > source venv/bin/activate
+
+   # Install the required dependencies into the venv
+   > pip3 install -r requirements.txt
+
+   # Build the documentation
+   > make html
```
Note for self: use Poetry or another automation solution to eliminate `make.sh`.

TheAssassin commented on source/conf.py:
```diff
@@ -201,15 +201,10 @@
-# Example configuration for intersphinx: refer to the Python standard library.
-intersphinx_mapping = {'https://docs.python.org/': None}
+# Intersphinx stub configuration
+intersphinx_mapping = {}
```
Just some feedback: I think intersphinx was once relevant because we planned to host separate docs repositories for different projects.

TheAssassin commented on source/conf.py:
```diff
@@ -124,6 +123,7 @@
 # relative to this directory. They are copied after the builtin static files,
 # so a file named "default.css" will overwrite the builtin "default.css".
 html_static_path = ['_static']
+html_css_files = ["css/custom.css"]
```
What version of Sphinx have you tried this with? I think in the past this setting didn't exist. I might be wrong on that, though.

Korne127 replied:
According to the [documentation](https://www.sphinx-doc.org/en/master/usage/configuration.html#confval-html_css_files), it has been added in version 1.8.
I don't know which exact version I used though right now, I'll assume the newest one as I just followed the requirements.txt file.

TheAssassin commented on source/contact.rst:
```diff
@@ -1,32 +1,25 @@
+------------------
 
-If you have further questions, please feel free to contact the AppImage team. The easiest and fastest way is to join our `IRC channel #appimage <ircs://irc.libera.chat/appimage>`__ on `Libera.Chat <https://libera.chat>`__ (`webchat <https://web.libera.chat/#AppImage>`__).
+An easy way to contact us is to ask questions on the `AppImage GitHub discussions page <https://github.com/orgs/AppImage/discussions>`__. Questions are usually answered in one to two days, and you'll get notified as soon as someone answered you. Posts stay, so you can always keep track of the conversation. You can also browse through other posts to see if someone else had the same question or pro ...
```
Please do not include what seems to be an SLA, we might not be able to keep up with such a standard. If you stay more vague (e.g., "a few days"), I could live with it.

Korne127 replied:
I wrote usually as this was how I assessed it when looking at the Github discussions (questions were often even answered in just a few hours), so rather as an estimate and not as a guarantee. I can change it though if you want.

TheAssassin commented on source/error-pages/404.rst:
```diff
@@ -1,17 +1,20 @@
 
 .. note::
-   As this documentation is still work in progress, the content might have been moved. Please use the search function or the table of contents to find its new location.
+   As this documentation is continuously being changed, the content might have been moved. Please use the search function or the table of contents to find its new location.
```
```suggestion
   As this documentation is continuously being updated, the content might have been moved. Please use the search function or the table of contents to find its new location.
```

Call it a marketing/PR wording improvement.

Korne127 replied:
Yeah, that sounds good :)

TheAssassin commented on source/index.rst:
```diff
@@ -14,19 +9,31 @@ Welcome to the AppImage documentation
+
+.. cssclass:: bold-link
+
+**If you are a first time user and just want to know how to run an AppImage (on your system or in a docker container) or have issues with an AppImage, see the** :ref:`Quickstart page <quickstart>`\ **!**
```
This sentence should contain less words IMO. People might read over the word "quickstart" more easily in the current state.

TheAssassin replied:
Also, don't you have to intend this for the statement above to have any effect?

Korne127 replied:
I can see your point. I feel like all of these are valid concerns (especially explicitly mentioning docker as the usual way to use (older) AppImages doesn't work on them so many might just look at the documentation for that).

I tried to make it something that people see and realize quickly by moving the big logo that took up all the space one initially looks at downloads and marking it bold so that it directly catches the eye.

What do you think about splitting it up into two statements like:

> **If you are a first time user and just want to run an AppImage, see the QuickStart page!**

> **If you want to run an AppImage in a docker container or have issues with an AppImage, also see the QuickStart page.**

Maybe I could also remove the "Welcome to the AppImage documentation." as it's duplicated with the headline anyways.

What do you mean with your second comment though?

But you're right, it's long.

TheAssassin commented on source/substitutions.rst:
```diff
@@ -0,0 +1,38 @@
+.. |appimage_preferred_source| replace:: In general, AppImages should be officially distributed by application authors. If application authors don't provide an AppImage, you should create an issue and ask them to start packaging the application as AppImage, or make a pull / merge request to add the creation of one if possible. Converting existing packages should only be used as a last resort if t ...
```
I'm not sure whether I like this substitution concept. Do you have any references on other docs that use such a concept? Overall, I think this can't easily be "converted" into Markdown, too. And I'd expect @probonopd to consider this very hard to maintain since without a real IDE, it's quite labor intensive to look these substitutions up (especially, if you're not aware of the feature in reST).

Korne127 replied:
The problem is that, with such a big documentation, there have to be several places with the same statements, e.g. as several categories / pages overlap or several pages might reference another page.

If you don't use substitutions, you end up with a lot of duplicates everywhere, and as long as you are not *very* careful to read through the entire documentation when you update one place, you might not catch everyone, leading to outdated information in some places and even information being inconsistent / contradicting.
I think that was a big part of why this documentation became harder to read, because often things were just updated in one place, missing others, so that there were many places without outdated information.

I understand your doubts, but I think that this will make the documentation much less prone to errors when updating it in the future, and also makes it easier to update is as you only have to modify this in one place but not look through the entire documentation.

TheAssassin commented on source/reference/appimage/reference-implementation.rst:
```diff
@@ -0,0 +1,111 @@
+AppImage reference implementation
+=================================
+
+The reference implementation of the :ref:`AppImage specification <appimage-specification>` is split up into several components, which are described on this page.
```
I'm not sure whether appimagetool is part of the reference implementation of the specification. The spec stands for itself and mostly describes the payload format but not tools that build AppImages. I have to think about this wording. The reference to the spec however is misleading in the current state IMO.

Korne127 replied:
The origin of this reference implementation section is the AppImageKit part of the former software overview page (which was in the introduction chapter). It described the runtime, appimagetool and additional helpers with a very similar structure to the now - reference implementation page.

Previously, the reference chapter just contained a section saying that the reference implementation of the AppImage specification is called "AppImageKit" and linked to the AppImageKit part on that page; therefore I thought that the reference implementation consists of all that (several statements on GitHub also seemed to define it that way).

If that's wrong and I should define the reference implementation as just the runtime and make an extra page for appimagetool instead, that's okay though.

Related to that, I noticed that there are some mistakes I made (and forgot to fix) while separating between separating between what the runtime does and what the appimagetool implementation does (see  https://github.com/AppImage/docs.appimage.org/pull/102#discussion_r1919918915). I'll do that soon and fix that up. :)

TheAssassin reviewed (commented):
Some feedback. I already see that reviewing this in-depth will take at least weeks. So far, it's mostly nitpicking, but then again, there's not a lot of content I already read.

Drsheppard01 commented:
If we're talking about this PR again, maybe somebody should include appimage specs as a separate page in an existing page instead of keeping it in a separate repository?

TheAssassin commented:
The separate repository makes maintaining it a lot easier. It should be referenced when needed. We plan to make releases (I think there's an open issue there), and we do not want to release the entire docs upon changes or block releases until the docs have been updated.

Drsheppard01 commented:
> We plan to make releases (I think there's an open issue there), and we do not want to release the entire docs upon changes or block releases until the docs have been updated.

But the documentation needs to change as the format changes, above you wrote about planning AppImage v3 and AppImage v4

Obviously AppImage Spec is logically part of the documentation it describes what AppImage is, and the documentation (docs.appimage.org) tells how to interact with it, and as the nature of the subject changes the way you interact with it changes, or as in the case of ubuntu/libfuse2 how to regain lost functionality

Korne127 commented:
@TheAssassin Hey, thank you very much for the answer and suggestions! I'm really happy that you noticed this PR and could take a look at it :)

Sorry for the late answer, I've had a lot to do the last few days and just wasn't on Github, but usually I respond pretty quickly here.

And yeah, I totally get that a real-time chat can help to discuss things, especially with such a big pull request. My matrix username is @korne127:matrix.org. :) Feel free to message me there.

probonopd commented:
> Obviously AppImage Spec is logically part of the documentation 

No, the spec is the spec (it describes the format in a formal, concise way). The documentation, in contrast, explains concept and gives examples on how to produce and consume AppImages (using tools).

Korne127 commented:
@TheAssassin Hey, I never received any message. I'd still be very happy to talk about this and get it merged; I think the documentation really deserves this upgrade. :)

probonopd commented:
@TheAssassin can you think of a way of rendering this for local reading in a formatted way before deploying? I'd like to read through everything in rendered form.

(Also, if we spend effort on the documentation, I'd strongly recommend to use Markdown as is done e.g., on https://hellosystem.github.io/docs/, but I guess that'll be a separate discussion.)

Korne127 commented:
@probonopd Look at my response in https://github.com/AppImage/docs.appimage.org/pull/102#issuecomment-2293670732 where I told how you can easily locally render it :)  

Iirc you yourself said that this works for you https://github.com/AppImage/docs.appimage.org/pull/102#issuecomment-2293692205 (and for the record, the tutorial on how to do this has also been added to the documentation in this PR). :)
